src/test/assembly/asm/global_asm.rs

```diff
@@ -24,3 +25,7 @@ global_asm!("movl ${}, %ecx", const 5, options(att_syntax));
 global_asm!("call {}", sym my_func);
 // CHECK: lea rax, [rip + MY_STATIC]
 global_asm!("lea rax, [rip + {}]", sym MY_STATIC);
+// CHECK: call _RNvCsiubXh4Yz005_10global_asm6foobar
+global_asm!("call {}", sym foobar);
+// CHECK: _RNvCsiubXh4Yz005_10global_asm6foobar:
+fn foobar() { loop {} }
```

Comment:
I started with an empty function here, and noticed an issue where `my_func` and `foobar` had been merged together by `mergefunc` pass, invalidating the reference from global asm. Probably `mergefunc` shouldn't touch symbols from `llvm.compiler.used`?

Reply:
No, mergefunc is perfectly valid here. In general Rust doesn't guarantee that two functions will have separate addresses.

Reply:
I mean, one of the functions is removed completely, and reference from the global asm is left unresolved.

Reply:
Oh right, that does seem like a bug. If something is in `llvm.compiler.used` then it's likely that it is being referenced using its symbol name, in which case mergefuncs shouldn't rename it. I *think* we can work around this by removing `unnamed_addr` from any symbols referenced by `llvm.compiler.used` in a pass after all values are generated.

Reply:
I think this is might be related to symbol export. 

Functions that are referenced from inline asm in text form should be listed in `exported_symbols`. We currently don't have any special treatment of asm_sym there.

Reply:
Could you give an example of what you have in mind?

As far as I know, for an inline assembly we do consider this aspect. For example, if a  function with an inline assembly block might be code generated in a different crate (because it is marked inline or generic) and it references a non-generic function, the referenced function will be exported as necessary, etc. In what situation would it be necessary in the case of global asm?

Reply:
asm_sym is injected into the assembly as text, so when doing LTO/ThinLTO, LLVM would consider the symbol to be unused. This is different from `llvm.compiler.used`, which only marks the function to be used, not the symbol.

Reply:
> This is different from `llvm.compiler.used`, which only marks the function to be used, not the symbol.

I see, but I think this is incorrect. Both `llvm.used` and `llvm.compiler.used` are about symbols, not function per se. The LLVM reference seems quite clear about that. Those lists are also limited to named values for that reason.

Reply:
I reread LLVM doc, you are correct about that. Some further digging shows that the issue is actually we generate an LLVM IR that marks the function as `internal`: https://godbolt.org/z/c4jPajT48 -- removing the internal linkage will make codegen correct. The symbol export actually doesn't affect the outcome directly -- but it is used indirectly from rustc_monomorphize to determine the linkage.